### DISCUSSION TITLE
Why not adding an argument to pass a custom model?

### DISCUSSION THREAD

**Discussion by AlexPasqua** [PFCCLab/PPOCRLabel#75] [Ideas]
Right now it seems it's necessary to modify the code if we want to use a custom model to perform auto-labelling. I think this is a pretty common use-case, so why not creating a couple of arguments that we can pass by command line to define what model we want to use? If None, then use default one.

If you agree, I can start a draft PR

**Comment by GreatV** [comment]
PR is wellcome

**Reply by AlexPasqua** [reply]
> PR is wellcome

It seems it was already implemented in #83

**Comment by GreatV** [comment]
The idea of adding an argument to pass a custom model for auto-labeling is indeed practical and aligns with common use cases, as highlighted in the discussion at [GitHub Discussion #75](https://github.com/PFCCLab/PPOCRLabel/discussions/75). Currently, as you mentioned, modifying the code is required to switch to a custom model. This can be cumbersome for users who frequently work with different models.

### Why Add Command-Line Arguments for Custom Models?
1. **Flexibility for Users:** Adding command-line arguments allows users to specify custom models without needing to modify the source code directly. This is particularly useful for those who experiment with or deploy multiple models.
2. **Common Use Case:** As auto-labeling often involves using pre-trained or custom-trained models, making it easier to integrate them would enhance the tool's usability.
3. **Code Maintainability:** Hardcoding paths or model configurations can lead to errors and make the codebase harder to maintain. Providing arguments for configuration encourages clean code practices.

### Suggested Implementation
To implement this feature, you could follow these steps:

1. **Add Command-Line Arguments:**
   Update the `argparse` section in the `get_main_app` function to include arguments for custom model paths. For example:
   ```python
   arg_parser.add_argument("--det_model_path", type=str, default=None, help="Path to custom detection model")
   arg_parser.add_argument("--rec_model_path", type=str, default=None, help="Path to custom recognition model")
   arg_parser.add_argument("--cls_model_path", type=str, default=None, help="Path to custom classification model")
   ```

2. **Modify the Model Initialization:**
   In the `MainWindow` class, modify the initialization of the `PaddleOCR` and `PPStructure` objects to use the provided model paths if specified. For example:
   ```python
   params = {
       "use_pdserving": False,
       "use_angle_cls": True,
       "det": True,
       "cls": True,
       "use_gpu": gpu,
       "lang": lang,
       "show_log": False,
   }

   if det_model_dir is not None:
       params["det_model_dir"] = det_model_dir
   if rec_model_dir is not None:
       params["rec_model_dir"] = rec_model_dir
   if rec_char_dict_path is not None:
       params["rec_char_dict_path"] = rec_char_dict_path
   if cls_model_dir is not None:
       params["cls_model_dir"] = cls_model_dir

   self.ocr = PaddleOCR(**params)
   ```

3. **Set Default Behavior:**
   If no custom model paths are provided, the application should fallback to the default models, ensuring backward compatibility.

### Draft Pull Request
If you decide to proceed, you can create a draft pull request (PR) with the proposed changes to gather input from the community and maintainers. Make sure to include:
- A description of the changes.
- Instructions for using the new command-line arguments.
- Details on how the changes maintain compatibility with existing functionality.

This feature would likely be welcomed by the community, as it simplifies the workflow for users who rely on custom models. You can reference [GitHub Discussion #75](https://github.com/PFCCLab/PPOCRLabel/discussions/75) in your PR to show that this enhancement addresses a discussed issue.

---
*Response generated by [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*
